element-handles-explainer.md

```diff
@@ -0,0 +1,418 @@
+
+The only reason this restriction exists is because a single handle can't refer to multiple elements. This may be a further argument to change `importhandles` to something like `importids`, and allow a single import could refer to multiple IDs.
+
+### Is there a need for a JavaScript object representation of a handle?
+
+There may be a need for a more advanced JavaScript API that allows interacting with element handles. It is not clear whether such an API is truly necessary. Are there good motivating examples that require this level of access to handles. Most cases should just be able to use the string representation of a handle.
+
+The goals of such an API would be:
+* Create a JavaScript object that represents `"host-element::handle(handle-name)"` in a way that does not break shadow DOM encapsulation (no direct access to the element that the handle refers to without having access to its shadow root). 
+* Work in a compatible way with properties of type `Element`, like [`Element.ariaActiveDescendantElement`](https://w3c.github.io/aria/#dom-ariamixin-ariaactivedescendantelement).
+
+One option is to have an `ElementHandle` simply be a mixin with `Element`, which adds a single property: `targetHandle` (readonly) the handle name as a string. The `targetHandle` property needs to be readonly, since changing the `targetHandle` would change the identity of the ElementHandle (change what element it refers to).
+
+Creating an element handle programmatically would require a special function like `document.createElementHandle(hostElement, handleName)`. 
+
+In that case, code that is unaware of `ElementHandle` would just see the host element. But if the code checks for `targetHandle`, it can work with it as appropriate.
+
+```html
+<input id="combobox-1" role="combobox" aria-activedescendant="x-listbox-1::handle(opt1)" />
+
+<x-listbox id="x-listbox-1">
+  #shadowRoot
+  | <div role="listbox" handle="the-listbox">
+  |   <div role="option" handle="opt1">Option 1</div>
+  |   <div role="option" handle="opt2">Option 2</div>
+  | </div>
+</x-listbox>
+
+<script>
+  const combobox = document.getElementById('combobox-1');
+
+  let activeDescendant = combobox.ariaActiveDescendantElement;
+  console.log(activeDescendant); // <x-listbox id="x-listbox-1">
+  console.log(activeDescendant.targetHandle); // 'opt1'
+
+  // If the shadowRoot is open, it is possible to get the actual element referenced by the handle:
+  const resolvedTarget = activeDescendant.shadowRoot?.getElementByHandle(activeDescendant.targetHandle);
+  console.log(resolvedTarget); // <div role="option" handle="opt1">Option 1</div>
+
+  // The attributes also allow an ElementHandle to be set on them
+  combobox.ariaActiveDescendantElement = document.createElementHandle(document.getElementById('x-listbox-1'), 'opt2');
+</script>
```

Comment:
In your original proposal, you had this concept of an "element handle" proxy object. I kind of liked it, but I admit that there isn't a strict need for it, since you can pass around strings instead. (Which is pretty similar to what developers do today in e.g. the React ecosystem.)

It does greatly simplify things that you don't need an extra kinda-sorta-Element proxy object, but on the other hand, authors will probably end up writing a lot of string concatenation and string parsing code instead. I wonder if it would make sense to introduce some helper APIs to do this? Or would that be overkill?

Reply:
The current proposal is an evolution of the proxy object idea from an earlier draft. Rather than an opaque object, it is instead the host Element itself, plus the handle name.

The reasons I abandoned the opaque proxy object were:
* It's not backwards compatible with APIs that return Element objects today, like `ariaActiveDescendantElement`. Existing code wouldn't know to check for the new object.
* I wanted to incorporate Alice's proposal to perform retargeting on the returned element, from [Encapsulation-preserving IDL Element reference attributes](https://github.com/WICG/aom/issues/195).
* The retargeted element is (usually) just the host element, so what if those attributes returned an "enhanced" retargeted element: something that was the host element _plus_ the handle to the element inside. You could use the handle to access the "real" target if desired (and the shadow root is open).

I don't think this proposal requires doing manual string parsing. The object is an Element (host), plus the (already-parsed) name of the handle.

Unless you're referring to an API similar to `getElementById`, but would take a string like `foo::handle(bar)` and return you the object representing foo plus the handle bar?

Reply:
Sorry for the late reply. I was gaming it out in my head, and it just seemed to me that I would probably end up doing string concatenation at some point to create strings `like::this(one)`.

For example, the `<x-listbox>` component may have a public JS API that returns the `handle` of the currently-active option, which means that the `<x-input>` component would have to construct the string `x-listbox::handle(<name here>)`. Maybe this is just a bad design, and the listbox component should use a consistent `handle` for whichever one happens to be active (as in your example), but there may be other cases where the outer component has to query the state of the inner component to figure out which kinds of handles are available.

Either way, it's not a big deal – it seems pretty straightforward to do the string concatenation.